pmap/pmap.go

```diff
@@ -0,0 +1,272 @@
+			port = p
+			break
+		}
+	}
+	return port
+}
+
+func stringInSlice(query string, list []string) bool {
+	for _, s := range list {
+		if s == query {
+			return true
+		}
+	}
+	return false
+}
+
+// NOTE: Unlike glusterd1's implementation, the search here is not overloaded
+// with delete operation. This is intentionally kept simple
+func registrySearch(brickname string, ptype PortType) int {
+	registry.RLock()
+	defer registry.RUnlock()
+
+	for p := registry.LastAlloc; p >= registry.BasePort; p-- {
+
+		if registry.Ports[p].Type != GfPmapPortFree {
+		}
+
+		if len(registry.Ports[p].Bricknames) == 0 || registry.Ports[p].Type != ptype {
+			continue
+		}
+
+		if stringInSlice(brickname, registry.Ports[p].Bricknames) {
+			return p
+		}
+	}
+
+	return 0
+}
+
+func registryAlloc(recheckForeign bool) int {
+	registry.Lock()
+	defer registry.Unlock()
+
+	var port int
+	for p := registry.BasePort; p <= gfPortMax; p++ {
+		if registry.Ports[p].Type == GfPmapPortFree ||
+			(recheckForeign && registry.Ports[p].Type == GfPmapPortForeign) {
+
+			if isPortFree(p) {
+				registry.Ports[p].Type = GfPmapPortLeased
+				port = p
+				break
+			} else {
+				// We may have an opportunity here to change
+				// the port's status from GfPmapPortFree to
+				// GfPmapPortForeign. Passing on for now...
+			}
+		}
+	}
+
+	if port > registry.LastAlloc {
+		registry.LastAlloc = port
+	}
+
+	return port
+}
+
+// AssignPort allocates and returns an available port. Optionally, if an
+// oldport specified for the brickpath, stale ports for the brickpath will
+// be cleaned up
+func AssignPort(oldport int, brickpath string) int {
+	if oldport != 0 {
+		registryRemove(0, brickpath, GfPmapPortBrickserver, nil)
+	}
+	return registryAlloc(true)
+}
+
+func registryBind(port int, brickname string, ptype PortType, xprt interface{}) {
+
+	if port > gfPortMax {
+		return
+	}
+
+	registry.Lock()
+	defer registry.Unlock()
+
+	registry.Ports[port].Type = ptype
+	registry.Ports[port].Bricknames = append(registry.Ports[port].Bricknames, brickname)
+	registry.Ports[port].Xprt = xprt
+
+	if registry.LastAlloc < port {
+		registry.LastAlloc = port
+	}
+}
+
+// opposite of append(), fast but doesn't maintain order
+func deleteFromSlice(list []string, query string) []string {
+
+	var found bool
+	var pos int
+	for i, s := range list {
+		if s == query {
+			pos = i
+			found = true
+			break
+		}
+	}
+
+	if found {
+		// swap i'th element with the last element
+		list[len(list)-1], list[pos] = list[pos], list[len(list)-1]
+		return list[:len(list)-1]
+	}
+
+	return list
+}
+
+func registryRemove(port int, brickname string, ptype PortType, xprt interface{}) {
+	if port > 0 {
+		if port > gfPortMax {
+			return
+		}
+	}
+
+	if brickname != "" {
+		port = registrySearch(brickname, ptype)
+		if port != 0 {
+			goto REMOVE
+		}
+	}
+
+	if xprt != nil {
+		port = registrySearchByXprt(xprt, ptype)
+		if port != 0 {
+			goto REMOVE
```

Comment:
Instead of using jumps, just make the REMOVE block a function and call it instead. Makes it cleaner.